src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Syntax/SyntaxTokenList.cs

```diff
@@ -0,0 +1,497 @@
+
+        // Add new tokens
+        foreach (var token in tokens)
+        {
+            builder.Add(token);
+        }
+
+        // If builder.Count == oldCount, there weren't any tokens added.
+        // So, there's no need to continue.
+        if (builder.Count == oldCount)
+        {
+            return this;
+        }
+
+        // Add remaining tokens starting from 'index'
+        for (var i = index; i < count; i++)
+        {
+            builder.Add(this[i]);
+        }
+
+        return new(parent: null, builder.ToGreenListNode(), position: 0, index: 0);
+    }
+
+    private SyntaxTokenList InsertRangeWithCount(int index, IEnumerable<SyntaxToken> tokens, int tokenCount)
+    {
+        if (tokenCount == 0)
+        {
+            return this;
+        }
+
+        var count = Count;
+        var array = new ArrayElement<GreenNode>[count + tokenCount];
+
+        // Add current tokens up to 'index'
+        int i;
+        for (i = 0; i < index; i++)
+        {
+            array[i].Value = this[i].RequiredNode;
+        }
+
+        // Add new tokens
+        foreach (var token in tokens)
+        {
+            array[i++].Value = token.RequiredNode;
+        }
+
+        // Add remaining tokens starting from 'index'
+        for (var j = index; j < count; i++, j++)
+        {
+            array[i].Value = this[j].RequiredNode;
+        }
+
+        return new(parent: null, InternalSyntax.SyntaxList.List(array), position: 0, index: 0);
+    }
+
+    public SyntaxTokenList RemoveAt(int index)
+    {
+        var count = Count;
+
+        ArgHelper.ThrowIfNegative(index);
+        ArgHelper.ThrowIfGreaterThanOrEqual(index, count);
+
+        // count - 1 because we're removing an item.
+        var array = new ArrayElement<GreenNode>[count - 1];
+
+        // Add current tokens up to 'index'
+        int i;
+        for (i = 0; i < index; i++)
+        {
+            array[i].Value = this[i].RequiredNode;
+        }
+
+        // Add remaining tokens starting *after* 'index'
+        for (var j = index + 1; j < count; i++, j++)
+        {
+            array[i].Value = this[j].RequiredNode;
+        }
+
+        return new(parent: null, InternalSyntax.SyntaxList.List(array), position: 0, index: 0);
+    }
+
+    public SyntaxTokenList Remove(SyntaxToken tokenInList)
+    {
+        var index = IndexOf(tokenInList);
+        return index >= 0 ? RemoveAt(index) : this;
+    }
+
+    public SyntaxTokenList Replace(SyntaxToken tokenInList, SyntaxToken newToken)
+    {
+        if (newToken == default)
+        {
+            ThrowHelper.ThrowArgumentOutOfRangeException(nameof(newToken));
+        }
+
+        return ReplaceRange(tokenInList, [newToken]);
+    }
+
+    public SyntaxTokenList ReplaceRange(SyntaxToken tokenInList, ReadOnlySpan<SyntaxToken> tokens)
+    {
+        var index = IndexOf(tokenInList);
+
+        if (tokens.Length == 0)
+        {
+            return RemoveAt(index);
+        }
+
+        if (index < 0)
```

Comment:
Feels like we should do this unconditionally, not avoid it if `tokens` is 0-length.